java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll

```diff
@@ -0,0 +1,167 @@
+ */
+
+import java
+import semmle.code.xml.MyBatisMapperXML
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.frameworks.MyBatis
+import semmle.code.java.frameworks.Properties
+
+private predicate propertiesKey(DataFlow::Node prop, string key) {
+  exists(MethodAccess m |
+    m.getMethod() instanceof PropertiesSetPropertyMethod and
+    key = m.getArgument(0).(CompileTimeConstantExpr).getStringValue() and
+    prop.asExpr() = m.getQualifier()
+  )
+}
+
+/** A data flow configuration tracing flow from ibatis obtaining the variable configuration object to setting the value of the variable. */
+private class PropertiesFlowConfig extends DataFlow2::Configuration {
+  PropertiesFlowConfig() { this = "PropertiesFlowConfig" }
+
+  override predicate isSource(DataFlow::Node src) {
+    exists(MethodAccess ma | ma.getMethod() instanceof IbatisConfigurationGetVariablesMethod |
+      src.asExpr() = ma
+    )
+  }
+
+  override predicate isSink(DataFlow::Node sink) { propertiesKey(sink, _) }
+}
+
+/** Get the key value of Mybatis Configuration Variable. */
+string getAMybatisConfigurationVariableKey() {
+  exists(PropertiesFlowConfig conf, DataFlow::Node n |
+    propertiesKey(n, result) and
+    conf.hasFlowTo(n)
+  )
+}
+
+/** A reference type that extends a parameterization of `java.util.List`. */
+class ListType extends RefType {
+  ListType() {
+    this.getSourceDeclaration().getASourceSupertype*().hasQualifiedName("java.util", "List")
+  }
+}
+
+/** Holds if the specified method uses MyBatis Mapper XMLElement. */
+predicate myBatisMapperXMLElementFromMethod(Method method, MyBatisMapperXMLElement mmxx) {
+  exists(MyBatisMapperSqlOperation mbmxe | mbmxe.getMapperMethod() = method |
+    mbmxe.getAChild*() = mmxx
+    or
+    exists(MyBatisMapperSql mbms |
+      mbmxe.getInclude().getRefid() = mbms.getId() and
+      mbms.getAChild*() = mmxx
+    )
+  )
+}
+
+/** Holds if the specified method uses Ibatis Sql Operation Annotation. */
+predicate myBatisSqlOperationAnnotationFromMethod(Method method, IbatisSqlOperationAnnotation isoa) {
+  exists(MyBatisSqlOperationAnnotationMethod msoam |
+    msoam = method and
+    msoam.getAnAnnotation() = isoa
+  )
+}
+
+/** Get the #{...} or ${...} parameters in the Mybatis mapper xml file. */
+string getAMybatisXmlSetValue(XMLElement xmle) {
+  result = xmle.getTextValue().regexpFind("(#|\\$)\\{[^\\}]*\\}", _, _)
+}
+
+/** Get the #{...} or ${...} parameters in the Mybatis sql operation annotation value. */
+string getAMybatisAnnotationSqlValue(IbatisSqlOperationAnnotation isoa) {
+  result = isoa.getSqlValue().regexpFind("(#|\\$)\\{[^\\}]*\\}", _, _)
+}
+
+/** Holds if it is SQL injection of MyBatis xml or MyBatis annotation. */
+bindingset[setValue]
+predicate isMybatisXmlOrAnnotationSqlInjection(
+  //DataFlow::Node node, MyBatisMapperXMLElement xmle, IbatisSqlOperationAnnotation isoa
+  DataFlow::Node node, MethodAccess ma, string setValue, MyBatisMapperXMLElement mmxe
```

Comment:
No, `mmxe` is only used in some disjuncts. That means in other disjuncts where it is not bound, this query will produce results for (node, ma, setValue) multiplied by every MyBatisMapperXMLElement in the program.

The case that uses `mmxe` must be lifted out of this function and only used in the XML case.